File: sections/100_grandmaster_finale/problems/05_dynamic_forest_ledger/solution.py
import sys

MOD = 1_000_000_007


class LinkCutTree:
    def __init__(self, values):
        n = len(values)
        self.child = [[0, 0] for _ in range(n + 1)]
        self.parent = [0] * (n + 1)
        self.reversed = [0] * (n + 1)
        self.size = [0] + [1] * n
        self.value = [0] + values
        self.total = [0] + values.copy()
        self.lazy_mul = [1] * (n + 1)
        self.lazy_add = [0] * (n + 1)

    def is_auxiliary_root(self, x):
        p = self.parent[x]
        return not p or (self.child[p][0] != x and self.child[p][1] != x)

    def pull(self, x):
        left, right = self.child[x]
        self.size[x] = 1 + self.size[left] + self.size[right]
        self.total[x] = (self.total[left] + self.value[x] + self.total[right]) % MOD

    def apply_reverse(self, x):
        if x:
            self.child[x].reverse()
            self.reversed[x] ^= 1

    def apply_affine(self, x, multiplier, addition):
        if not x:
            return
        self.value[x] = (multiplier * self.value[x] + addition) % MOD
        self.total[x] = (multiplier * self.total[x] + addition * self.size[x]) % MOD
        self.lazy_mul[x] = multiplier * self.lazy_mul[x] % MOD
        self.lazy_add[x] = (multiplier * self.lazy_add[x] + addition) % MOD

    def push(self, x):
        if self.reversed[x]:
            self.apply_reverse(self.child[x][0])
            self.apply_reverse(self.child[x][1])
            self.reversed[x] = 0
        if self.lazy_mul[x] != 1 or self.lazy_add[x]:
            multiplier, addition = self.lazy_mul[x], self.lazy_add[x]
            self.apply_affine(self.child[x][0], multiplier, addition)
            self.apply_affine(self.child[x][1], multiplier, addition)
            self.lazy_mul[x] = 1
            self.lazy_add[x] = 0

    def rotate(self, x):
        p = self.parent[x]
        g = self.parent[p]
        side = self.child[p][1] == x
        middle = self.child[x][side ^ 1]
        if not self.is_auxiliary_root(p):
            self.child[g][self.child[g][1] == p] = x
        self.parent[x] = g
        self.child[x][side ^ 1] = p
        self.parent[p] = x
        self.child[p][side] = middle
        if middle:
            self.parent[middle] = p
        self.pull(p)
        self.pull(x)

    def splay(self, x):
        ancestors = [x]
        y = x
        while not self.is_auxiliary_root(y):
            y = self.parent[y]
            ancestors.append(y)
        for y in reversed(ancestors):
            self.push(y)
        while not self.is_auxiliary_root(x):
            p = self.parent[x]
            g = self.parent[p]
            if not self.is_auxiliary_root(p):
                self.rotate(p if ((self.child[p][1] == x)
                                  == (self.child[g][1] == p)) else x)
            self.rotate(x)

    def access(self, x):
        previous = 0
        current = x
        while current:
            self.splay(current)
            self.child[current][1] = previous
            self.pull(current)
            previous = current
            current = self.parent[current]
        self.splay(x)

    def make_root(self, x):
        self.access(x)
        self.apply_reverse(x)

    def link(self, u, v):
        self.make_root(u)
        self.parent[u] = v

    def cut(self, u, v):
        self.make_root(u)
        self.access(v)
        self.child[v][0] = 0
        self.parent[u] = 0
        self.pull(v)

    def expose_path(self, u, v):
        self.make_root(u)
        self.access(v)
        return v


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n, q = map(int, data[:2])
    tree = LinkCutTree(list(map(int, data[2:2 + n])))
    index = 2 + n
    output = []
    for _ in range(q):
        operation = data[index]
        u, v = int(data[index + 1]), int(data[index + 2])
        index += 3
        if operation == b"LINK":
            tree.link(u, v)
        elif operation == b"CUT":
            tree.cut(u, v)
        elif operation == b"AFFINE":
            multiplier, addition = int(data[index]), int(data[index + 1])
            index += 2
            tree.apply_affine(tree.expose_path(u, v), multiplier, addition)
        else:
            output.append(str(tree.total[tree.expose_path(u, v)]))
    print("\n".join(output))


if __name__ == "__main__":
    main()

File: sections/100_grandmaster_finale/problems/05_dynamic_forest_ledger/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

constexpr int64 MOD = 1'000'000'007;

struct LinkCutTree {
    int n;
    vector<array<int, 2>> child;
    vector<int> parent, reversed, size;
    vector<int64> value, sum, lazy_mul, lazy_add;

    explicit LinkCutTree(const vector<int64> &initial)
        : n(initial.size()), child(n + 1), parent(n + 1), reversed(n + 1),
          size(n + 1, 1), value(n + 1), sum(n + 1), lazy_mul(n + 1, 1),
          lazy_add(n + 1) {
        size[0] = 0;
        for (int i = 1; i <= n; ++i) value[i] = sum[i] = initial[i - 1];
    }

    bool is_auxiliary_root(int x) const {
        int p = parent[x];
        return p == 0 || (child[p][0] != x && child[p][1] != x);
    }

    void pull(int x) {
        int left = child[x][0], right = child[x][1];
        size[x] = 1 + size[left] + size[right];
        sum[x] = (sum[left] + value[x] + sum[right]) % MOD;
    }

    void apply_reverse(int x) {
        if (x == 0) return;
        swap(child[x][0], child[x][1]);
        reversed[x] ^= 1;
    }

    void apply_affine(int x, int64 multiplier, int64 addition) {
        if (x == 0) return;
        value[x] = (multiplier * value[x] + addition) % MOD;
        sum[x] = (multiplier * sum[x] + addition * size[x]) % MOD;
        lazy_mul[x] = multiplier * lazy_mul[x] % MOD;
        lazy_add[x] = (multiplier * lazy_add[x] + addition) % MOD;
    }

    void push(int x) {
        if (reversed[x]) {
            apply_reverse(child[x][0]);
            apply_reverse(child[x][1]);
            reversed[x] = 0;
        }
        if (lazy_mul[x] != 1 || lazy_add[x] != 0) {
            apply_affine(child[x][0], lazy_mul[x], lazy_add[x]);
            apply_affine(child[x][1], lazy_mul[x], lazy_add[x]);
            lazy_mul[x] = 1;
            lazy_add[x] = 0;
        }
    }

    void rotate(int x) {
        int p = parent[x], g = parent[p];
        int side = child[p][1] == x;
        int middle = child[x][side ^ 1];
        if (!is_auxiliary_root(p)) child[g][child[g][1] == p] = x;
        parent[x] = g;
        child[x][side ^ 1] = p;
        parent[p] = x;
        child[p][side] = middle;
        if (middle) parent[middle] = p;
        pull(p);
        pull(x);
    }

    void splay(int x) {
        vector<int> ancestors{x};
        for (int y = x; !is_auxiliary_root(y);) {
            y = parent[y];
            ancestors.push_back(y);
        }
        while (!ancestors.empty()) {
            push(ancestors.back());
            ancestors.pop_back();
        }
        while (!is_auxiliary_root(x)) {
            int p = parent[x], g = parent[p];
            if (!is_auxiliary_root(p))
                rotate((child[p][1] == x) == (child[g][1] == p) ? p : x);
            rotate(x);
        }
    }

    void access(int x) {
        int previous = 0;
        for (int current = x; current; current = parent[current]) {
            splay(current);
            child[current][1] = previous;
            pull(current);
            previous = current;
        }
        splay(x);
    }

    void make_root(int x) {
        access(x);
        apply_reverse(x);
    }

    void link(int u, int v) {
        make_root(u);
        parent[u] = v;
    }

    void cut(int u, int v) {
        make_root(u);
        access(v);
        child[v][0] = 0;
        parent[u] = 0;
        pull(v);
    }

    int expose_path(int u, int v) {
        make_root(u);
        access(v);
        return v;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;
    vector<int64> initial(n);
    for (int64 &x : initial) cin >> x;
    LinkCutTree tree(initial);
    while (q--) {
        string operation;
        int u, v;
        cin >> operation >> u >> v;
        if (operation == "LINK") {
            tree.link(u, v);
        } else if (operation == "CUT") {
            tree.cut(u, v);
        } else if (operation == "AFFINE") {
            int64 multiplier, addition;
            cin >> multiplier >> addition;
            tree.apply_affine(tree.expose_path(u, v), multiplier, addition);
        } else {
            cout << tree.sum[tree.expose_path(u, v)] << '\n';
        }
    }
}
